Purchases/Identity/RCPurchaserInfoManager.m

```diff
@@ -0,0 +1,196 @@
+    }];
+}
+
+- (void)fetchAndCachePurchaserInfoIfStaleWithAppUserID:(NSString *)appUserID
+                                     isAppBackgrounded:(BOOL)isAppBackgrounded
+                                            completion:(nullable RCReceivePurchaserInfoBlock)completion {
+    RCPurchaserInfo * _Nullable cachedPurchaserInfo = [self cachedPurchaserInfoForAppUserID:appUserID];
+    BOOL isCacheStale = [self.deviceCache isPurchaserInfoCacheStaleForAppUserID:appUserID
+                                                              isAppBackgrounded:isAppBackgrounded];
+    BOOL needsToRefresh = isCacheStale || !cachedPurchaserInfo;
+    if (needsToRefresh) {
+        RCDebugLog(@"%@", isAppBackgrounded
+                          ? RCStrings.purchaserInfo.purchaserinfo_stale_updating_in_background
+                          : RCStrings.purchaserInfo.purchaserinfo_stale_updating_in_foreground);
+
+        [self fetchAndCachePurchaserInfoWithAppUserID:appUserID
+                                    isAppBackgrounded:isAppBackgrounded
+                                           completion:completion];
+        RCSuccessLog(@"%@", RCStrings.purchaserInfo.purchaserinfo_updated_from_network);
+    } else {
+        if (completion) {
+            [self.operationDispatcher dispatchOnMainThread: ^{
+                completion(cachedPurchaserInfo, nil);
+            }];
+        }
+    }
+}
+
+- (void)sendUpdatedPurchaserInfoToDelegateIfChanged:(RCPurchaserInfo *)info {
+
+    if ([self.delegate respondsToSelector:@selector(purchaserInfoManagerDidReceiveUpdatedPurchaserInfo:)]) {
+        @synchronized (self) {
+            if (![self.lastSentPurchaserInfo isEqual:info]) {
+                if (self.lastSentPurchaserInfo) {
+                    RCDebugLog(@"%@", RCStrings.purchaserInfo.sending_updated_purchaserinfo_to_delegate);
+                } else {
+                    RCDebugLog(@"%@", RCStrings.purchaserInfo.sending_latest_purchaserinfo_to_delegate);
+                }
+                self.lastSentPurchaserInfo = info;
+                [self.operationDispatcher dispatchOnMainThread:^{
+                    [self.delegate purchaserInfoManagerDidReceiveUpdatedPurchaserInfo:info];
+                }];
+            }
+        }
+    }
+}
+
+- (void)purchaserInfoWithAppUserID:(NSString *)appUserID
+                   completionBlock:(nullable RCReceivePurchaserInfoBlock)completion {
+    RCPurchaserInfo * _Nullable infoFromCache = [self cachedPurchaserInfoForAppUserID:appUserID];
+    if (infoFromCache) {
+        RCDebugLog(@"%@", RCStrings.purchaserInfo.vending_cache);
+        if (completion) {
+            [self.operationDispatcher dispatchOnMainThread: ^{
+                completion(infoFromCache, nil);
+            }];
+        }
+    }
+
+    [self.systemInfo isApplicationBackgroundedWithCompletion:^(BOOL isAppBackgrounded) {
+        BOOL shouldCallCompletionAfterFetching = infoFromCache == nil && completion != nil;
```

Comment:
I had some trouble understanding this part. I didn't understand why you were not just passing completion to `fetchAndCachePurchaserInfoIfStaleWithAppUserID` even if it's nil. I now get that it's to avoid the completion being called again if it was called already in line 167, but it took me a bit.

I was thinking maybe the check for `[self.systemInfo isApplicationBackgroundedWithCompletion` could be done in `fetchAndCachePurchaserInfoIfStaleWithAppUserID`. Then calling `fetchAndCachePurchaserInfoIfStaleWithAppUserID` in line 169 passing a nil completion.

This is just a suggestion including my personal preference, it is not necessarily the way I think this method should be written. If you think your current implementation is clear, leave it as is :)

Reply:
Agreed, this looks pretty confusing

I did some cleanup here: I added an explicit boolean to store whether we've already called completion, and added a comment. lmk if it feels better! Doing the isAppBackgrounded check in `fetchAndCachePurchaserInfoIfStaleWithAppUserID` is a slightly larger refactor, but I'm happy to do it if the new logic is still confusing